Add OpenSSH Proxy Jump docs (#13170)
gravitational/teleport: docs/pages/server-access/guides/openssh.mdx

</Details>

> <Details title="Using Trusted Clusters?">

Author: > It looks like the -J flag applies to the [tsh command as a whole](https://github.com/gravitational/teleport/blob/2aae5b58992dda6fcfde27ead967029a52fe098f/tool/tsh/tsh.go#L508), not just tsh proxy ssh. Should docs re: the -J flag be in a more visible location? This guide is intended for infrastructure with legacy OpenSSH servers, and it looks like the -J flag is more generally applicable.

The main intention of the proxy templates change is to `[improve] tsh proxy command interoperability with ssh client and connecting to leaf cluster proxies directly` - https://github.com/gravitational/teleport/blob/master/rfd/0062-tsh-proxy-template.md.
While the `-J` flag is usable through other commands like `tsh ssh -J {{proxy}}`, I believe the only use case we explicitly support is `tsh proxy ssh -J {{proxy}}`, which is specifically intended for use in theS SSH config's `ProxyCommand` field.

So in my opinion the OpenSSH guide is the best 

fit for the current implementation which is geared towards using the OpenSSH client to connect to OpenSSH hosts and Teleport nodes through the Teleport Auth/Proxy servers.

> It seems like there haven't been any changes to the [template used in writeSSHConfig](https://github.com/gravitational/teleport/blob/2aae5b58992dda6fcfde27ead967029a52fe098f/tool/tsh/config.go#L34-L45). Should we be instructing users to run tsh config, then manually edit the resulting config to include the ProxyCommand with the -J flag as shown below? Otherwise, I'm not sure this guide is the best place to put the -J flag instructions.

The current template is still the best default template for users to use. To use Proxy templates, they'd need to edit the config and proxy templates themselves to fit their naming conventions etc. This guide is intended for users who want to customize their ssh config beyond the default config we provide with `tsh config`.
 
To me it makes sense to provide these "next step" instructions right here. 

I'll try adding some more context in the guide, but if you have an idea for a more fitting spot in the docs we can try that.

> <Details title="Using Trusted Clusters?">

Reviewer: Thanks for clarifying! I think the best way to go would be to edit the text into a more explicit set of steps. We should tell the reader to edit the config generated by `tsh config` to use the `{{proxy}}` template, then add a `proxy_templates` field to the `tsh` config file. Right now, these sound more like general possibilities, and it can be clearer how they fit into the workflow we describe in this guide.

client configuration block for the root cluster and all currently known leaf clusters.

This configuration block uses a `ProxyCommand` directive to connect to SSH 
servers in a leaf cluster through the root cluster's Proxy Service.

In scenarios where Proxy Service instances in a leaf cluster are reachable by
SSH clients, you might prefer to connect directly through the leaf proxies to
achieve lower latency.

To support this scenario, the `tsh proxy ssh` command provides the `--jumphost`/`-J` flag.
This flag points to the public address of a cluster's Proxy Service and indicates 
that the OpenSSH client will be connecting through that proxy instead of the proxy
you're currently logged into.

```txt
Host *.{{ .NodeName }}.{{ .ClusterName }}
   Port 3022
   ProxyCommand tsh proxy ssh -J proxy.leaf1.example.com:443 %r@%h:%p
```

When the `-J` flag is used with the `{{proxy}}` template variable, the jump
proxy address and host server address can be dynamically determined using proxy
templates defined in the `tsh` configuration file (`~/.tsh/config/config.yaml`
or a global `/etc/tsh.yaml`)

```txt
Host *.example.com
   Port 3022
   ProxyCommand tsh proxy ssh -J {{proxy}} %r@%h:%p
```

> ```yaml

Reviewer: This YAML snippet needs some introduction. I would add a paragraph that transitions from the last snippet. For example:

```markdown
To configure the way `tsh` evaluates the `{{proxy}}` template variable when
using the `-J` flag, add a `proxy_templates` field to your `tsh` configuration
file (`~/.tsh/config/config.yaml` or `/etc/tsh.yaml` for all users on your system):
```
We could then delete the mention of the `tsh` configuration file location from above.

Again, if this configuration only applies to the `-J` flag, I'm guessing we need some explicit instructions for the user to edit their `tsh config`-created configuration to include the `-J` flag. Also, we should probably introduce the `-J` flag outside of this guide, since it looks like it's intended for all `tsh` commands, not just those involving legacy OpenSSH servers.

Author: > Also, we should probably introduce the `-J` flag outside of this guide, since it looks like it's intended for all `tsh` commands, not just those involving legacy OpenSSH servers.

We currently mention the `-J` flag in one other spot - https://goteleport.com/docs/ver/10.0/server-access/guides/tsh/#ssh-jump-host

And then there are there's the ssh man pages which are relevant as well:

```
     -J destination
             Connect to the target host by first making a ssh connection to the jump host described by destination and then establishing
             a TCP forwarding to the ultimate destination from there.  Multiple jump hops may be specified separated by comma charac‐
             ters.  This is a shortcut to specify a ProxyJump configuration directive.  Note that configuration directives supplied on
             the command-line generally apply to the destination host and not any specified jump hosts.  Use ~/.ssh/config to specify
             configuration for jump hosts.
```

Do you think we need to expand the tsh jump host section? If so, how will we fit in these docs to include the OpenSSH client integration with `tsh config`? Maybe that's where an `OpenSSH Client guide` would come in?

</Details>

<Details title="Experiencing latency in leaf cluster Nodes?">

### Proxy Jump

In the generated OpenSSH client configuration, the `ProxyCommand` for each leaf 
cluster connects through the root cluster's Proxy Service. In scenarios where
leaf cluster Proxy Services are reachable by SSH client, you might prefer to

> connect directly through the leaf proxies to achieve lower latency.

Reviewer 2: Nit:
```suggestion
connect directly through the leaf proxies for lower latency.
```
Otherwise it's "achieve" again in the next sentence.

> To achieve this, update the `ProxyCommand` of the leaf cluster's configuration

Reviewer: ```suggestion
To enable direct connections to a Proxy Service in a leaf cluster, open the 
SSH configuration file you generated earlier and update
the `ProxyCommand` of the leaf cluster's configuration
```

I would specify that this configuration block is in the SSH config the reader generated earlier in this guide. If a reader is following the guide, the configuration file they generated won't be in the default location.

block to use the leaf Proxy Service as a jumphost, using the `-J` flag.

```txt
Host *.{{ .NodeName }}.leaf1.example.com
   Port 3022
   ProxyCommand tsh proxy ssh -J proxy.leaf1.example.com:443 %r@%h:%p
```

### Proxy Templates

> To streamline this customization, you can configure Proxy Templates, which

Reviewer: "this customization" can be more specific. Maybe replace this paragraph with:

````markdown
With Proxy Templates, `tsh` will dynamically determine the address of the
Proxy Service to connect to based on the address of the destination
host in your `ssh` command.
````

can be used with `tsh proxy ssh -J {{proxy}}` to dynamically determine the
proxy address based on the host address you are connecting to.

> To use proxy templating, add `-J {{proxy}}` to the `ProxyCommand` line in

Reviewer: ```suggestion
To use Proxy Templates, add `-J {{proxy}}` to the `ProxyCommand` line in
```
For consistency.

your `~/.ssh/config`.

```txt
Host *.example.com
   Port 3022
   ProxyCommand tsh proxy ssh -J {{proxy}} %r@%h:%p
```

Then, add `proxy_templates` to your `tsh` configuration file (`~/.tsh/config/config.yaml`
or a global `/etc/tsh.yaml`).

```yaml
proxy_templates:
- template: '^(\w+)\.(leaf1\.example\.com):([0-9]+)$'
  proxy: "$2:443"
```

`tsh proxy ssh -J {{proxy}}` will attempt to match the host server address `%h:%p` with the 
configured templates. If there is a match, then the jump proxy address `{{proxy}}` will
be replaced using the template's `proxy` field and the host server address `%h:%p` will be 
replaced using the template's `host` field if set.

| Field      | Description |
| ---------- | ----------- |
| `template` | (Required) Regular expression that the host server address `%h:%p` is matched against. |
| `proxy`    | (Required) Proxy Service address to use for proxy jump. Can reference capturing groups from the regular expression in `template` (e.g., `$1` or `$2`). |

> | `host`     | (Optional) Host Servera ddress to connect to. Can reference capturing groups from the regular expression in `template` (e.g., `$1` or `$2`). Defaults to full host spec `%h:%p`. |

Reviewer 2: ```suggestion
| `host`     | (Optional) Host Server address to connect to. Can reference capturing groups from the regular expression in `template` (e.g., `$1` or `$2`). Defaults to full host spec `%h:%p`. |
```

</Details>

> <Details title="Using Proxy Jump in Trusted Clusters?">

Reviewer: ```suggestion
<Details title="Using Trusted Clusters?">
```
Suggested because `tsh config` generates the config with a `ProxyCommand` directive, so this question would always be false for users who have been following the instructions. I think this Details box would be relevant for anyone using Trusted Clusters.

> Normally, you connect to leaf cluster SSH nodes through the root cluster proxy,

Reviewer: I think we can start this paragraph by referring to the SSH config generated above, rather than what normally happens—I'm not sure what "normally" means in this context. Maybe something like:

```markdown
The configuration file that `tsh config` generates uses a `ProxyCommand`
directive to connect to SSH servers in a leaf cluster through the root cluster's
Proxy Service.
```

which is provided in the `ProxyCommand` line of the SSH config generated above.

> In scenarios where leaf proxies are reachable by SSH clients, you might prefer

Reviewer: ```suggestion
In scenarios where Proxy Service instances in a leaf cluster are reachable by SSH clients, you might prefer
```

> to connect directly through the leaf proxies to avoid potential latency issues

Reviewer: I think we can make "to avoid potential latency issues of the connection being routed through the root cluster" a bit more concise by replacing it with "to achieve lower latency".

of the connection being routed through the root cluster.

To support this scenario, the `tsh proxy ssh` command provides the `-J` flag.

> This flag points at a cluster's web proxy address and indicates that the

Reviewer: ```suggestion
This flag points to the public address of a cluster's Proxy Service and indicates that the
```

OpenSSH client will be connecting through that proxy instead of the proxy
you're currently logged into:

```txt
Host *.{{ .NodeName }}.{{ .ClusterName }}
   Port 3022
   ProxyCommand tsh proxy ssh -J proxy.leaf1.example.com:443 %r@%h:%p
```

> The jump proxy address and the node name to connect to can be dynamically

Reviewer: ```suggestion
The `-J` flag dynamically determines the jump proxy address and the name of the host to connect to
```

determined from the SSH config's full hostname spec (`%h:%p`) using `{{proxy}}`
template variable:

```txt
Host *.example.com
   Port 3022
   ProxyCommand tsh proxy ssh -J {{proxy}} %r@%h:%p
```

The rules of how to parse the proxy address and node name from the full host
spec can be defined in the `tsh` configuration file (`~/.tsh/config/config.yaml`
or a global `/etc/tsh.yaml`):

```yaml
proxy_templates:
- template: '^(\w+)\.(leaf1\.example\.com):(.*)$'
  proxy: "$2:443"
- template: '^(\w+)\.(leaf2\.example\.com):(.*)$'
  proxy: "$2:3080"
  host: "$1"
```

| Field      | Description |
| ---------- | ----------- |

> | `template` | (Required) Regular expression full SSH config host spec is matched against. |

Author: I've added a sentence above

| `proxy`    | (Required) Web proxy address to use. Can reference regex groups from `template`. |

> | `host`     | (Optional) Node name to use. Can reference regex groups from `template`. Defaults to full host spec. |

Author: This should make more sense with the sentence above

> Given the configuration above, the following command will connect to the node

Reviewer: I think we should use "host" instead of "node" when talking about servers that are running `openssh` but not Teleport. This way, there's no risk of confusion with Teleport Nodes.

Author: Actually in this example these are only Teleport nodes. I am adding an example of connecting to an openSSH host though.

`node-1.leaf1.example.com:3022` through the proxy `leaf1.example.com:443`:

```code
$ ssh -p 3022 root@node-1.leaf1.example.com
```

> The following command will connect to the node `node-1:3022` through the proxy

Author: node-1 is a Teleport node, not an openssh host

</Details>

> <Details title="Using Trusted Clusters?">

Reviewer: Looking at this guide now, it seems a bit awkward that the paragraph before the Details boxes begins, "If you are using Trusted Clusters...". Maybe we can remove the "If you are using Trusted Clusters..." paragraph and begin this Details box with something like,

```markdown
If you are using Trusted Clusters, the `tsh config` command will print an OpenSSH
 client configuration block for the root cluster and all currently known leaf clusters.

This configuration block uses a `ProxyCommand` directive to connect to SSH 
servers in a leaf cluster through the root cluster's Proxy Service.
```

The configuration file that `tsh config` generates uses a `ProxyCommand`
directive to connect to SSH servers in a leaf cluster through the root cluster's
Proxy Service.

In scenarios where Proxy Service instances in a leaf cluster are reachable by
SSH clients, you might prefer to connect directly through the leaf proxies to
achieve lower latency.

To support this scenario, the `tsh proxy ssh` command provides the `-J` flag.
This flag points to the public address of a cluster's Proxy Service and indicates 
that the OpenSSH client will be connecting through that proxy instead of the proxy
you're currently logged into:

```txt
Host *.{{ .NodeName }}.{{ .ClusterName }}
   Port 3022
   ProxyCommand tsh proxy ssh -J proxy.leaf1.example.com:443 %r@%h:%p
```

The `-J` flag dynamically determines the jump proxy address and the name
of the host to connect to using the SSH config's full hostname spec (`%h:%p`)
and the `{{proxy}}` template variable:

The jump proxy address and the node name to connect to can be dynamically
determined from the SSH config's full hostname spec (`%h:%p`) using `{{proxy}}`

> template variable:

Reviewer: This paragraph duplicates the one above.

```txt
Host *.example.com
   Port 3022
   ProxyCommand tsh proxy ssh -J {{proxy}} %r@%h:%p
```

> You can define the rules of how to parse the proxy address and node name

Reviewer: ```suggestion
You can define rules for how to parse the Proxy Service address and Node name
```

> from the full host in the `tsh` configuration file (`~/.tsh/config/config.yaml`

Reviewer: Is "full host" the same as "hostname provided in an `ssh` command"?

Author: No, it's referring to `%h:%p`, and the ssh `Hostname` is just the `%h`. I've changed the wording to "host server address" for clarity.

> or a global `/etc/tsh.yaml`). `tsh proxy ssh` will take the attempt to match the

Reviewer: ```suggestion
or a global `/etc/tsh.yaml`). `tsh proxy ssh` will take the attempt to match the
```
```suggestion
or a global `/etc/tsh.yaml`). 

`tsh proxy ssh` will attempt to match the
```

full hostname `%h:%p` with one of the provided templates, and use that template
to replace `{{proxy}}` and optionally `%h:%p` with the desired values.

```yaml
proxy_templates:
- template: '^(\w+)\.(leaf1\.example\.com):([0-9]+)$'
  proxy: "$2:443"
- template: '^(\w+)\.(leaf2\.example\.com):([0-9]+)$'
  proxy: "$2:3080"
  host: "$1:$3"
- template: '(\w+(\.\w+)*)\.(example\.com):([0-9]+)$'
  proxy: "leaf1.example.com:443"
  host: "$1:22"
```

| Field      | Description |
| ---------- | ----------- |

> | `template` | (Required) Regular expression full SSH config host spec `%h:%p` is matched against. |

Reviewer: ```suggestion
| `template` | (Required) Regular expression that the full SSH config host spec `%h:%p` is matched against. |
```

| `proxy`    | (Required) Proxy Service address to use. Can reference capturing groups from the regular expression in `template` (e.g., `$1` or `$2`). |
| `host`     | (Optional) Full hostname to connect to. Can reference capturing groups from the regular expression in `template` (e.g., `$1` or `$2`). Defaults to full host spec `%h:%p`. |

> Given the configuration above, the following command will connect to the node

Reviewer: ```suggestion
Given the configuration above, the following command will connect to the Node
```

`node-1.leaf1.example.com:3022` through the Proxy Service `leaf1.example.com:443`:

```code
$ ssh root@node-1.leaf1.example.com
```

> The following command will connect to the node `node-1:3022` through the Proxy Service

Reviewer: ```suggestion
The following command will connect to the Node `node-1:3022` through the Proxy Service
```
